sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,744 @@
+  az_span remainder = az_span_copy(destination_span, mqtt_url_prefix);
+  remainder = az_span_copy(remainder, az_span_slice(iot_hub, 0, iot_hub_length));
+  remainder = az_span_copy(remainder, mqtt_url_suffix);
+  az_span_copy_u8(remainder, null_terminator);
+
+  return AZ_OK;
+}
+
+// Read the user environment variables used to connect to IoT Hub
+static az_result read_configuration_and_init_client()
+{
+  az_span cert = AZ_SPAN_FROM_BUFFER(x509_cert_pem_file);
+  AZ_RETURN_IF_FAILED(
+      read_configuration_entry(ENV_DEVICE_X509_CERT_PEM_FILE, NULL, false, cert, &cert));
+
+  az_span trusted = AZ_SPAN_FROM_BUFFER(x509_trust_pem_file);
+  AZ_RETURN_IF_FAILED(
+      read_configuration_entry(ENV_DEVICE_X509_TRUST_PEM_FILE, "", false, trusted, &trusted));
+
+  az_span device_id_span = AZ_SPAN_FROM_BUFFER(device_id);
+  AZ_RETURN_IF_FAILED(
+      read_configuration_entry(ENV_DEVICE_ID, "", false, device_id_span, &device_id_span));
+
+  az_span iot_hub_hostname_span = AZ_SPAN_FROM_BUFFER(iot_hub_hostname);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      ENV_IOT_HUB_HOSTNAME, "", false, iot_hub_hostname_span, &iot_hub_hostname_span));
+
+  // Paho requires that the MQTT endpoint be of the form ssl://<HUB ENDPOINT>:8883
+  AZ_RETURN_IF_FAILED(
+      create_mqtt_endpoint(mqtt_endpoint, (int32_t)sizeof(mqtt_endpoint), iot_hub_hostname_span));
+
+  // Initialize the hub client with the hub host endpoint and the default connection options
+  AZ_RETURN_IF_FAILED(az_iot_hub_client_init(
+      &client,
+      az_span_slice(iot_hub_hostname_span, 0, (int32_t)strlen(iot_hub_hostname)),
+      az_span_slice(device_id_span, 0, (int32_t)strlen(device_id)),
+      NULL));
+
+  return AZ_OK;
+}
+
+static az_result report_method(az_span payload, az_span response, az_span* out_response)
+{
+  az_result result;
+
+  // Parse the "since" field in the payload
+  az_json_parser jp;
+  az_json_token_member tm;
+  AZ_RETURN_IF_FAILED(az_json_parser_init(&jp, payload));
+  AZ_RETURN_IF_FAILED(az_json_parser_parse_token(&jp, &tm.token));
+  if (tm.token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  az_span start_time_span = AZ_SPAN_NULL;
+  while (az_succeeded(az_json_parser_parse_token_member(&jp, &tm)))
+  {
+    if (az_span_is_content_equal(method_payload_value, tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm.token, &start_time_span));
+    }
+
+    // else ignore token.
+  }
+
+  // Get the current time as a string
+  time_t rawtime;
+  struct tm* timeinfo;
+  time(&rawtime);
+  timeinfo = localtime(&rawtime);
+  size_t len = strftime(end_time_buffer, sizeof(end_time_buffer), "%Y-%m-%dT%H:%M:%S%z", timeinfo);
+  az_span time_span = az_span_init((uint8_t*)end_time_buffer, (int32_t)len);
+
+  // Build the method response payload
+  az_json_builder json_builder;
+  result = az_json_builder_init(&json_builder, response, NULL);
```

Comment:
humm, result is not handled?

Reply:
changed to wrapped in `AZ_RETURN_IF_FAILED`